igvc_navigation/include/igvc_navigation/Graph.h

```diff
@@ -0,0 +1,275 @@
+/**
+The graph contains the nodes and cells that set the foundation of the search
+problem.
+
+In Field D*, Nodes are located on the corners of grid cells (as opposed to the
+center, like in A*, D*, or D*lite).
+
+The Graph object provides an interface for the occupancy grid, as well as nodes.
+To facilitate the search problem, the graph operates under a unit-cell assumption,
+meaning that each cell has side lengths of 1m. The `Resolution` parameter specifies
+the conversion between the unit cell and the actual cell dimensions. i.e. if Resolution
+is 0.2 then each cell actually has side dimensions of 0.2m.
+
+Author: Alejandro Escontrela <aescontrela3@gatech.edu>
+Date Created: December 16, 2018
+*/
+
+#ifndef GRAPH_H
+#define GRAPH_H
+
+// #define NDEBUG # TODO: uncomment for production
+
+#include "igvc_utils/NodeUtils.hpp"
+#include "Node.h"
+
+#include <igvc_msgs/map.h>
+
+#include <cv_bridge/cv_bridge.h>
+#include <utility>
+#include <vector>
+#include <tuple>
+#include <cmath>
+#include <limits>
+#include <assert.h>
+
+class Graph
+{
+public:
+    /**
+    Prev map is the occupancy grid one timestep ago, it's used to check for changes in the
+    the occupancy grid. Map is the current, most up-to-date occupancy grid.
+    */
+    cv_bridge::CvImagePtr Map;
+
+    Node Start; // start node in the search problem
+    Node Goal; // goal node in the search problem
+
+    // Updated cell information is used to update nodes that lie on the 4 corners of
+    // each updated cell. This is reset each time updateGraph is called. Each element
+    // is composed of <cell index, old value, new value>
+    std::vector<std::tuple<int,int>> updatedCells;
```

Comment:
Updated the documentation to better reflect functionality. I'm leaving it as a tuple just in case more dimensions need to be incorporated in the future. Also, I don't think a float tuple carries significantly more overhead than a pair.

Reply:
I think a struct would be more helpful here to attach meaning to the different values, and it also can be easily modified

Reply:
I think a struct is overkill here. A vector of `std::tuple<float,float>` can be extended by adding more fields to the tuple. The vector only contain the indices of updated cells in the graph search problem, which can be inferred from the variable name `updatedCells`.

Reply:
I don't think it is overkill, since a pair is essentially a struct, so if a struct was overkill, then by transitivity, a pair would also be overkill. The difference between the two is that, as @barulicm has said, the struct and it's members are named, which provides greater clarity to your code.

Reply:
`std::tuple` and `std::pair` are two different data structures; this code implements `std::tuple`, not `std::pair`. Anyways, let's say that I create a `Cell` struct. The only two fields in this struct would be `x` and `y`. This would hardly add any meaning given that the code is documented to reflect that each `std::tuple` in the vector is an <x,y> index for the cell.

Reply:
Semantic meaning expressed directly in the code is better than documentation comments.

tuple and pair aren’t all that different, and pair is largely regarded as a specialization of tuple. 

Structs would also help make it more obvious which parts of the code are handling the same data. There are a few spots in this code where there is a vector of tuples, structs would make it clearer if those vectors contained the same type or not.

Finally, “overkill” as measured by what? A struct is a very light and nimble tool which predates C and was created with the explicit purpose of grouping multiple data members together for convenient collective handling.

Reply:
I agree with creating a `Cell` struct to differentiate between the data types contained within the various vectors in the code. I'll define it in `Graph.h`